### ISSUE TITLE
Received disconnect: Timeout, your session not responding.

### ISSUE THREAD

**Issue by eozer** [ronf/asyncssh#425]
Hi,

I am trying to use `asyncssh` on top of [python_prompt_toolkit](https://github.com/prompt-toolkit/python-prompt-toolkit) to modernize a legacy console (colorize output, autocompletion _etc._), which can be connected via SSH. I want to set an interactive input and use the same session, so that users can use the existing console with added features. The code below does the job okay, until users do not send any command (_i.e_, session stays idle) after a few seconds (around 20-30 seconds). Then the protocol error is received: "Timeout, your session not responding"; connection and channel are closed.

Here is the code:

```python
#!/usr/bin/env python

import asyncio
import logging
import sys

import asyncssh
from asyncssh.connection import SSHClientConnection
from asyncssh.process import SSHClientProcess


from prompt_toolkit.shortcuts import print_formatted_text
from prompt_toolkit.shortcuts.prompt import PromptSession


logging.basicConfig()
logger = logging.getLogger()
logger.setLevel(logging.DEBUG)
asyncssh.logger.set_debug_level(3)

# Alias 'print_formatted_text', so that 'print' calls go to the SSH client.
print = print_formatted_text
PROMPT = "< "


async def wait_until_banner(p: SSHClientProcess) -> str:
    out = ""
    while True:
        out = await asyncio.wait_for(p.stdout.read(65536), timeout=5)
        print(out, end='')
        if PROMPT in out:
            break
    return out


async def interact() -> None:
    prompt_session = PromptSession(message="")
    async with asyncssh.connect(host='1.2.3.4', username="foo", password="bar", keepalive_interval=5000) as conn:
        conn: SSHClientConnection
        async with conn.create_process(term_type="auto") as p:
            p: SSHClientProcess
            await wait_until_banner(p)
            # Interactive user input loop
            while True:
                out = ""
                cmd = ""
                try:
                    cmd = await asyncio.wait_for(prompt_session.prompt_async(), timeout=5)
                    p.stdin.write(f"{cmd}\r\n")
                    while True:
                        try:
                            out = await asyncio.wait_for(p.stdout.read(65536), timeout=15)
                            print(out, end='')
                            if "DESTROYING SESSION" in out and "EXIT;" in cmd:
                                return
                            elif PROMPT in out:
                                break
                        except:
                            conn.logger.debug("Timeout")
                            break
                except:
                    pass  # conn.send_packet(98) ?


try:
    asyncio.get_event_loop().run_until_complete(interact())
except (OSError, asyncssh.Error) as exc:
    sys.exit('SSH connection failed: ' + str(exc))

```

Here is the output:
```
...snip until prompt
<  
DEBUG:asyncssh:[conn=0, chan=0, pktid=13] Received MSG_CHANNEL_REQUEST (98), 31 bytes
  00000000: 62 00 00 00 00 00 00 00 15 6b 65 65 70 61 6c 69  b........keepali
  00000010: 76 65 40 6f 70 65 6e 73 73 68 2e 63 6f 6d 01     ve@openssh.com.
DEBUG:asyncssh:[conn=0, chan=0] Received OpenSSH keepalive channel request
DEBUG:asyncssh:[conn=0, pktid=15] Sent MSG_IGNORE (2), 5 bytes
  00000000: 02 00 00 00 00                                   .....
DEBUG:asyncssh:[conn=0, chan=0, pktid=16] Sent MSG_CHANNEL_SUCCESS (99), 5 bytes
  00000000: 63 00 00 00 00                                   c....
DEBUG:asyncssh:[conn=0, pktid=14] Received MSG_UNIMPLEMENTED (3), 5 bytes
  00000000: 03 00 00 00 10                                   .....
 
DEBUG:asyncssh:[conn=0, chan=0, pktid=15] Received MSG_CHANNEL_REQUEST (98), 31 bytes
  00000000: 62 00 00 00 00 00 00 00 15 6b 65 65 70 61 6c 69  b........keepali
  00000010: 76 65 40 6f 70 65 6e 73 73 68 2e 63 6f 6d 01     ve@openssh.com.
DEBUG:asyncssh:[conn=0, chan=0] Received OpenSSH keepalive channel request
DEBUG:asyncssh:[conn=0, pktid=17] Sent MSG_IGNORE (2), 5 bytes
  00000000: 02 00 00 00 00                                   .....
DEBUG:asyncssh:[conn=0, chan=0, pktid=18] Sent MSG_CHANNEL_SUCCESS (99), 5 bytes
  00000000: 63 00 00 00 00                                   c....
DEBUG:asyncssh:[conn=0, pktid=16] Received MSG_UNIMPLEMENTED (3), 5 bytes
  00000000: 03 00 00 00 12                                   .....
 
DEBUG:asyncssh:[conn=0, chan=0, pktid=17] Received MSG_CHANNEL_REQUEST (98), 31 bytes
  00000000: 62 00 00 00 00 00 00 00 15 6b 65 65 70 61 6c 69  b........keepali
  00000010: 76 65 40 6f 70 65 6e 73 73 68 2e 63 6f 6d 01     ve@openssh.com.
DEBUG:asyncssh:[conn=0, chan=0] Received OpenSSH keepalive channel request
DEBUG:asyncssh:[conn=0, pktid=19] Sent MSG_IGNORE (2), 5 bytes
  00000000: 02 00 00 00 00                                   .....
DEBUG:asyncssh:[conn=0, chan=0, pktid=20] Sent MSG_CHANNEL_SUCCESS (99), 5 bytes
  00000000: 63 00 00 00 00                                   c....
DEBUG:asyncssh:[conn=0, pktid=18] Received MSG_UNIMPLEMENTED (3), 5 bytes
  00000000: 03 00 00 00 14                                   .....
 
DEBUG:asyncssh:[conn=0, pktid=19] Received MSG_DISCONNECT (1), 50 bytes
  00000000: 01 00 00 00 02 00 00 00 25 54 69 6d 65 6f 75 74  ........%Timeout
  00000010: 2c 20 79 6f 75 72 20 73 65 73 73 69 6f 6e 20 6e  , your session n
  00000020: 6f 74 20 72 65 73 70 6f 6e 64 69 6e 67 2e 00 00  ot responding...
  00000030: 00 00                                            ..
DEBUG:asyncssh:[conn=0] Received disconnect: Timeout, your session not responding. (2)
INFO:asyncssh:[conn=0] Connection failure: Timeout, your session not responding.
INFO:asyncssh:[conn=0, chan=0] Closing channel due to connection close
INFO:asyncssh:[conn=0, chan=0] Channel closed: Timeout, your session not responding.
```

On Windows SSH to the same server and staying idle (as long as the server allows) does not disconnect. Here, I could see from debug lines that the server sends packet type 98 [SSH_MSG_CHANNEL_REQUEST](https://www.iana.org/assignments/ssh-parameters/ssh-parameters.xhtml), and client sends back packet type 100 [SSH_MSG_CHANNEL_FAILURE](https://www.iana.org/assignments/ssh-parameters/ssh-parameters.xhtml). I am not sure if this is what keeps the session alive.

```
PS: C:\ssh -V
OpenSSH_for_Windows_8.1p1, LibreSSL 3.0.2

PS: C:\ssh <problematic_server_IP> -vvv
... snip until prompt
< debug3: receive packet: type 98
debug1: client_input_channel_req: channel 0 rtype keepalive@openssh.com reply 1
debug3: send packet: type 100
debug3: receive packet: type 98
debug1: client_input_channel_req: channel 0 rtype keepalive@openssh.com reply 1
debug3: send packet: type 100
...
```

Another weird thing is that the python code works okay when connected to a linux server. That is, client sessions kept alive as long as the server allows. So there is something else I need to do to keep the session alive for the problematic server.

Do you have any pointer to help me?

**Comment by ronf** [comment]
Do you know what kind of SSH server you're having this problem with?

From the look of things, the server is sending an OpenSSH-style keepalive message, but it is not expecting to get back MSG_CHANNEL_SUCCESS. In fact, it actually responds with MSG_UNIMPLEMENTED when it gets that reply, and apparently also doesn't treat that success response as an acknowledgement to its keepalive, causing it to eventually time out even though a keepalive response is being sent.

Your Windows SSH client doesn't understand these keepalive messages, so it returns MSG_FAILURE. Generally speaking, either a success or a failure response to the keepalive@openssh.com channel request is usually treated by the sender as a sign the connection should be kept alive. However, in this case, it seems to be accepting the failure response but not the success response. If you try the same test with an OpenSSH client to the problematic server, what do you see?

**Comment by eozer** [comment]
I will check what kind of SSH server is running, for now unfortunately I do not know.

Actually, Windows OpenSSH client was connected to the problematic server (see the last code section). Or do you mean something else actually?

**Comment by ronf** [comment]
Ah, sorry - I missed that the Windows client you were using was also OpenSSH.

I rechecked the keepalive implementation of OpenSSH, and it looks like OpenSSH as a client does respond to channel requests for keepalive@openssh.com with a failure (packet type 100), where AsyncSSH currently responds with success (packet type 99). The version of OpenSSH server I'm running here accepts either as a valid keepalive response, though. There's also an extra MSG_IGNORE in there in some cases. For instance:

```
debug2: channel 0: request keepalive@openssh.com confirm 1
debug3: send packet: type 98
debug3: receive packet: type 2
debug3: Received SSH2_MSG_IGNORE
debug3: receive packet: type 99
debug1: Got 99/18 for keepalive
```

I wonder if the SSH server you are having a problem with is expecting only a failure response to this request. I could probably change AsyncSSH to send a failure response here, but since OpenSSH accepted both it seemed cleaner to me to send success to indicate that the keepalive messages where explicitly understood.

**Comment by eozer** [comment]
I could also try to send failure response to verify it is the case. I realized that I made a mistake in the first code section "pass  # conn.send_packet(98) ?" line. It should have been instead conn.send_packet(100). I do not have access to the problematic server at the moment, but I suppose this could be tried first from my part.

How about from AsyncSSH point of view though? Is it the function below from https://github.com/ronf/asyncssh/blob/master/asyncssh/channel.py#L390, responsible for sending keepalive requests?

```python
def _report_response(self, result):
    """Report back the response to a previously issued channel request"""

    request, _, want_reply = self._request_queue.pop(0)

    if want_reply and self._send_state not in {'close_pending', 'closed'}:
        if result:
            self.send_packet(MSG_CHANNEL_SUCCESS)
        else:
            self.send_packet(MSG_CHANNEL_FAILURE)
```

**Comment by ronf** [comment]
No - it's handled at https://github.com/ronf/asyncssh/blob/master/asyncssh/channel.py#L600.

```python
    def _process_keepalive_at_openssh_dot_com_request(self, packet):
        """Process an incoming OpenSSH keepalive request"""

        packet.check_end()

        self.logger.debug2('Received OpenSSH keepalive channel request')
        return True
```

If this function didn't exist, the function which processes channel requests would automatically return MSG_CHANNEL_FAILURE when it got a keepalive@openssh.com channel request. Since the function exists and returns `True`, though, that gets translated into a MSG_CHANNEL_SUCCESS. You could also leave the function but have it return `False` instead of `True`. That way, you still get the benefit of the additional debugging message.

**Comment by ronf** [comment]
Regarding the "conn.send_packet()" calls, what are you trying to accomplish there? You shouldn't be sending standalone MSG_CHANNEL_SUCCESS or MSG_CHANNEL_FAILURE messages. They are only valid as responses to incoming channel requests, and they must contain the channel number. That might explain why you are getting "UNIMPLEMENETED" errors. Also, if you start sending extra reply messages like this, you could cause other legitimate responses to not be associated with the right requests.

**Comment by eozer** [comment]
Okay thanks for the explanations. I can try return `False` from `_process_keepalive_at_openssh_dot_com_request`.

**Comment by eozer** [comment]
Server version is SSH-2.0-OpenSSH_4.3-40020723001f00.
```
PS: C:\ssh <problematic_server_IP> -vvv
...
debug1: Local version string SSH-2.0-OpenSSH_for_Windows_8.1
debug1: Remote protocol version 2.0, remote software version OpenSSH_4.3-40020723001f00
debug1: match: OpenSSH_4.3-40020723001f00 pat OpenSSH_2*,OpenSSH_3*,OpenSSH_4* compat 0x00000002
...
```

I can verify also, returning `False` from `_process_keepalive_at_openssh_dot_com_request`, i.e, responding MSG_CHANNEL_FAILURE, indeed keeps the session alive for my case as well. Unfortunately, I need to study AsyncSSH a little bit more to send a pull request with a proper solution.

**Comment by ronf** [comment]
Wow - OpenSSH 4.3 is REALLY old. It was released over 15 years ago (Feb 1, 2006).

I think you may be running into https://bugzilla.mindrot.org/show_bug.cgi?id=1307, which was fixed in OpenSSH back in version 4.9 in 2008, or perhaps the change to accept more packet types was an accidental side effect of another keepalive-related change (and I see a handful of those over the years).

That said, if sending a MSG_CHANNEL_FAILURE instead of MSG_CHANNEL_SUCCESS in AsyncSSH is enough to work around the problem, I'd be willing to make the change to `_process_keepalive_at_openssh_dot_com_request` to do that, since modern versions of OpenSSH seems to accept either. There's no need to submit a pull request here -- I can take care of this after I get some other changes checked in. I'm hoping it's just a one-line change.

**Comment by eozer** [comment]
Yes the server has a very old OpenSSH version. I wish I could just update it, but... It is just a black box for me at this point.

Thank you @ronf for your support! I think we can close this issue. It is okay from my point of view.

**Comment by ronf** [comment]
Understood about not being able to upgrade the server - that's another reason why I'm happy to try and provide a fix here, as long as I can keep it simple.

Feel free to leave this open for now, and I'll close it once I have the change available. In the meantime, I'm glad you're able to work around the problem!

**Comment by ronf** [comment]
This is now in the "develop" branch as commit f568610 and will be included in the next release.

**Comment by ronf** [comment]
This change is now available in AsyncSSH 2.9.0.
